pkg/protocol/tipmanager/v1/tipmanager.go

```diff
@@ -0,0 +1,363 @@
+
+		return true
+	})
+
+	return tipSet
+}
+
+// Evict evicts a slot from the TipManager.
+func (t *TipManager) Evict(slotIndex iotago.SlotIndex) {
+	if t.markSlotAsEvicted(slotIndex) {
+		if evictedObjects, deleted := t.tipMetadataStorage.DeleteAndReturn(slotIndex); deleted {
+			evictedObjects.ForEach(func(_ iotago.BlockID, tipMetadata *TipMetadata) bool {
+				tipMetadata.evicted.Trigger()
+
+				return true
+			})
+		}
+	}
+}
+
+// Events returns the events of the TipManager.
+func (t *TipManager) Events() *tipmanager.Events {
+	return t.events
+}
+
+// Shutdown marks the TipManager as shutdown.
+func (t *TipManager) Shutdown() {
+	t.TriggerStopped()
+}
+
+// setupBlockMetadata sets up the behavior of the given Block.
+func (t *TipManager) setupBlockMetadata(tipMetadata *TipMetadata) {
+	tipMetadata.stronglyConnectedToTips.OnUpdate(func(_, isConnected bool) {
+		t.updateParents(tipMetadata, propagateConnectedChildren(isConnected, true))
+	})
+
+	tipMetadata.weaklyConnectedToTips.OnUpdate(func(_, isConnected bool) {
+		t.updateParents(tipMetadata, propagateConnectedChildren(isConnected, false))
+	})
+
+	tipMetadata.OnIsStrongTipUpdated(func(isStrongTip bool) {
+		if isStrongTip {
+			t.strongTipSet.Set(tipMetadata.Block().ID(), tipMetadata)
+		} else {
+			t.strongTipSet.Delete(tipMetadata.Block().ID())
+		}
+	})
+
+	tipMetadata.OnIsWeakTipUpdated(func(isWeakTip bool) {
+		if isWeakTip {
+			t.weakTipSet.Set(tipMetadata.Block().ID(), tipMetadata)
+		} else {
+			t.weakTipSet.Delete(tipMetadata.Block().ID())
+		}
+	})
+
+	tipMetadata.OnEvicted(func() {
+		t.strongTipSet.Delete(tipMetadata.Block().ID())
+		t.weakTipSet.Delete(tipMetadata.Block().ID())
+	})
+
+	tipMetadata.setTipPool(t.determineTipPool(tipMetadata))
+
+	t.events.BlockAdded.Trigger(tipMetadata)
+}
+
+// determineTipPool determines the initial TipPool of the given Block.
+func (t *TipManager) determineTipPool(tipMetadata *TipMetadata, minPool ...tipmanager.TipPool) tipmanager.TipPool {
+	if lo.First(minPool) <= tipmanager.StrongTipPool && !t.conflictDAG.AcceptanceState(tipMetadata.Block().ConflictIDs()).IsRejected() {
+		return tipmanager.StrongTipPool
+	}
+
+	if lo.First(minPool) <= tipmanager.WeakTipPool && t.conflictDAG.LikedInstead(tipMetadata.Block().PayloadConflictIDs()).Size() == 0 {
+		return tipmanager.WeakTipPool
+	}
+
+	return tipmanager.DroppedTipPool
+}
+
+// updateParents updates the parents of the given Block.
+func (t *TipManager) updateParents(tipMetadata *TipMetadata, updates map[model.ParentsType]func(*TipMetadata)) {
+	if parentBlock := tipMetadata.Block(); parentBlock != nil && parentBlock.Block() != nil {
+		parentBlock.ForEachParent(func(parent model.Parent) {
+			metadataStorage := t.metadataStorage(parent.ID.Index())
+			if metadataStorage == nil {
+				return
+			}
+
+			parentMetadata, created := metadataStorage.GetOrCreate(parent.ID, func() *TipMetadata { return NewBlockMetadata(lo.Return1(t.retrieveBlock(parent.ID))) })
```

Comment:
Ideally, the entire BlockDAG / tangle should follow the promise based paradigm and by the time we arrive at the tip manager, the parents should be well known and directly reachable / propagated in the corresponding metadata objects.

The promised based approach would populate them as part of the solidification process, but since we don't have this paradigm, yet and there are no direct references to the parents anywhere in the metadata, I thought it would be a simple solution to simply request the missing parents "on the fly" from the cache (ignoring missing evicted parents that we don't want to add to our tip pool anyway).

This will btw. be revisited in a future PR, where we introduce "orphanage" as an additional flag in the tip manager that will come with the introduction of TSC and this will be a parent based property where we need to set up behavior depending on our parents anyway and here it sounds like it might make sense to actually get and populate the parents a single time first, but I would like to address that in a future PR.